### PR TITLE
feat: add C implementation for `math/base/special/gamma`

### PR DESCRIPTION
**Pull request by gunjjoshi**
## Description

> What is the purpose of this pull request?

This pull request:

-   adds C implementation for [`math/base/special/gamma`](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/gamma).

```c
double stdlib_base_gamma( const double x );
```

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves a part of #649.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/math/base/special/gamma/src/main.c [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+		return 0.0;
+	}
+	q = stdlib_base_abs( x );
+	if ( q > 33.0 ) {
+		if ( x >= 0.0 ) {
+			return stirlingApprox( x );
+		}
+		p = stdlib_base_floor( q );
+		// Check whether `x` is even...
+		i = p;
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/gamma/src/main.c]:
```suggestion
		i = stdlib_base_floor( q );
		// Check whether `x` is even...
```

File: lib/node_modules/@stdlib/math/base/special/gamma/src/main.c [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+		if ( ( i & 1 ) == 0 ) {
+			sign = -1.0;
+		} else {
+			sign = 1.0;
+		}
+		z = q - p;
```
Comment (Pranavchiku) [inline lib/node_modules/@stdlib/math/base/special/gamma/src/main.c:218]:
```suggestion
		p = i;
		z = q - p;
```

**Review by Pranavchiku — APPROVED** [review]
This PR looks good, once the changes are done this will be ready to merge. Also the above changes are minor, I am not exactly sure which has higher precision but I think the suggested one has.  Here is my discussion with chatGPT, https://chatgpt.com/share/d247d6fe-6321-49b7-a469-a92d0ab38b18

File: lib/node_modules/@stdlib/math/base/special/gamma/src/main.c (hunk 1 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/gamma/src/main.c]: @Pranavchiku Not clear why your changes are superior to the current code. What is missing here is explicit casting. E.g.,

```c
i = (int32_t)p;
```

Casting should (almost) always be made explicit in order to reinforce for the reader what type a receiving value is.

Additionally, just as in JS, `// comments` should always be proceeded by a blank line when the preceding line is an expression. E.g.,

```c
p = stdlib_base_floor( q );

// Check whether `x` is even...
```

File: lib/node_modules/@stdlib/math/base/special/gamma/src/main.c (hunk 2 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/gamma/src/main.c:218]: This doesn't seem necessary and could result in precision loss compared to the original code. What does the original upstream implementation do?

File: lib/node_modules/@stdlib/math/base/special/gamma/src/main.c (hunk 1 above)
Comment (gunjjoshi) [reply lib/node_modules/@stdlib/math/base/special/gamma/src/main.c]: Thanks for the suggestions @kgryte and @Pranavchiku. I have made the changes now.

File: lib/node_modules/@stdlib/math/base/special/gamma/src/main.c (hunk 2 above)
Comment (gunjjoshi) [reply lib/node_modules/@stdlib/math/base/special/gamma/src/main.c:218]: In the original `cephes` implementation, we have this:

```c
i = p;
if( (i & 1) == 0 )
	sgngam = -1;
else
	sgngam = 1;
z = q - p;
```

File: lib/node_modules/@stdlib/math/base/special/gamma/manifest.json [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        "@stdlib/math/base/special/abs",
+        "@stdlib/math/base/special/floor",
+        "@stdlib/math/base/special/sin",
+        "@stdlib/constants/float64/pinf",
+        "@stdlib/constants/float64/ninf",
+        "@stdlib/constants/float64/pi",
+        "@stdlib/constants/float64/eulergamma",
+        "@stdlib/constants/float64/sqrt-two-pi",
+        "@stdlib/math/base/special/pow",
+        "@stdlib/math/base/special/exp"
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/gamma/manifest.json]: `pinf` is listed twice:

```suggestion
        "@stdlib/math/base/napi/unary",
        "@stdlib/math/base/assert/is-nan",
        "@stdlib/math/base/assert/is-integer",
        "@stdlib/math/base/assert/is-negative-zero",
        "@stdlib/math/base/special/abs",
        "@stdlib/math/base/special/floor",
        "@stdlib/math/base/special/sin",
        "@stdlib/constants/float64/pinf",
        "@stdlib/constants/float64/ninf",
        "@stdlib/constants/float64/pi",
        "@stdlib/constants/float64/eulergamma",
        "@stdlib/constants/float64/sqrt-two-pi",
        "@stdlib/math/base/special/pow",
        "@stdlib/math/base/special/exp"
```

File: lib/node_modules/@stdlib/math/base/special/gamma/manifest.json [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/gamma/manifest.json]:
```suggestion
        "@stdlib/math/base/assert/is-nan",
        "@stdlib/math/base/assert/is-integer",
        "@stdlib/math/base/assert/is-negative-zero",
        "@stdlib/math/base/special/abs",
        "@stdlib/math/base/special/floor",
        "@stdlib/math/base/special/sin",
        "@stdlib/constants/float64/pinf",
        "@stdlib/constants/float64/ninf",
        "@stdlib/constants/float64/pi",
        "@stdlib/constants/float64/eulergamma",
        "@stdlib/constants/float64/sqrt-two-pi",
        "@stdlib/math/base/special/pow",
        "@stdlib/math/base/special/exp"
```

File: lib/node_modules/@stdlib/math/base/special/gamma/manifest.json [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/gamma/manifest.json]:
```suggestion
        "@stdlib/math/base/assert/is-nan",
        "@stdlib/math/base/assert/is-integer",
        "@stdlib/math/base/assert/is-negative-zero",
        "@stdlib/math/base/special/abs",
        "@stdlib/math/base/special/floor",
        "@stdlib/math/base/special/sin",
        "@stdlib/constants/float64/pinf",
        "@stdlib/constants/float64/ninf",
        "@stdlib/constants/float64/pi",
        "@stdlib/constants/float64/eulergamma",
        "@stdlib/constants/float64/sqrt-two-pi",
        "@stdlib/math/base/special/pow",
        "@stdlib/math/base/special/exp"
```

File: lib/node_modules/@stdlib/math/base/special/gamma/src/main.c [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+// END: rational_pq
+
+/* End auto-generated functions. */
+
+/**
+* Evaluates the gamma function using Stirling's formula. The polynomial is valid for \\(33 \leq x \leq 172\\).
+*
+* @private
+* @param x    input value
+* @returns    function value
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/gamma/src/main.c]:
```suggestion
* @param x    input value
* @return     function value
```

File: lib/node_modules/@stdlib/math/base/special/gamma/src/main.c [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+	return STDLIB_CONSTANT_FLOAT64_SQRT_TWO_PI * y * w;
+}
+
+/**
+* Evaluates the gamma function using a small-value approximation.
+*
+* @private
+* @param x    input value
+* @param z    scale factor
+* @returns    function value
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/math/base/special/gamma/src/main.c]:
```suggestion
* @param x    input value
* @param z    scale factor
* @return     function value
```

**Review by Planeshifter — APPROVED** [review]
Let's land this! 🚀

**Pull request merged**